src/charm.py

```diff
@@ -18,3 +28,65 @@ def __init__(self, *args):
             args: Arguments passed to the CharmBase parent constructor.
         """
         super().__init__(*args)
+        self._charm_state = None
+        self._saml_integrator = None
+        self.framework.observe(self.on.config_changed, self._on_config_changed)
+        self.framework.observe(self.on.upgrade_charm, self._on_upgrade_charm)
+        self.framework.observe(self.on.saml_relation_created, self._on_saml_relation_created)
```

Comment:
As commented for `saml.py` This should be listened in the library and the charm should listen to internal events emitted by the library

Reply:
The library will be added in a follow up PR. I'm resolving this comment for now